### PR TITLE
Fixes #2 how to serve static assets

### LINKED ISSUE #2: Dockerflow apps shouldn't have to serve static files.
There isn't a good story around the serving of static files that are part of a Dockerflow app.

### PR DESCRIPTION
**Pull request by mostlygeek**
@chuckharmston r? please
@lmorchard r? as well since this will also affect TestPilot

This is the initial draft on how Dockerflow application should serve static assets.
TL;DR: from the application and cached with a reverse proxy for performance.

### CODE REVIEW

**Comment by lmorchard** [comment]
This sounds good to me, especially since it's how we're already doing it on Test Pilot - except for the bit about source files in `.dockerignore`, which also sounds good but we're not doing it yet.

FWIW, we're using a package called [Whitenoise](http://whitenoise.evans.io/en/stable/), which claims to be very fast & efficient at serving up static assets from a Python app. If we need better performance than what it offers, we can look into using it as origin for a CDN.

File: docs/serving-static-content.md [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+
+## Generated static assets
+
+When using transpiling tools like [bablejs](https://babeljs.io/), [sass](http://sass-lang.com/) and [browserify](http://browserify.org/) only have their output and not the source in the container. Use a [.dockerignore](https://github.com/mozilla/testpilot/blob/master/.dockerignore) file to exclude source directories. If source materials are required it is OK to include them in the container to be used at run time. 
+
+## How to serve static content
+
+### Serve it with the application
+
+This is the recommended approach. It is the most idiomatic usage of containers to hide application dependencies from the host machine. For Dockerflow applications the docker host only needs to provide a `$PORT` and need not be concerned about what's inside.
```
Comment (chuckharmston) [inline docs/serving-static-content.md]: I don't think it's always safe to assume that apps can, should, or will be good at serving their own static files. Many applications will apply a bunch of middleware and other overhead on every request, and will just be outright inefficient when acting as a fileserver proxy.

**Comment by chuckharmston** [comment]
This covers some use cases (such as ours) fine, but if we're looking to standardize this across cloudops, I think additional options will be required. Two possibilities:

- Allow application developers to create additional static file containers that use nginx as a reverse proxy, similar to our [compose](https://github.com/mozilla/universal-search-recommendation/blob/master/docker-compose.yml#L12-L18) [setup](https://github.com/mozilla/universal-search-recommendation/tree/master/conf/nginx).
- A build process that includes compiling static files to a predetermined directory and pushing them to S3.

File: docs/serving-static-content.md (hunk 1 above)
Comment (mostlygeek) [inline docs/serving-static-content.md]: Safe or fast? 

If safe, as in security, then the application must be secure. Putting the app within the container isolates it from the host machines file system. Though we definitely don't want it to serve a config file with secrets. Apps must limit access to what static files they serve. 

If fast, having the nginx caching proxy in front will make this fast. As fast as serving the files off disk directly. This means the application should add appropriate `Cache-Control` headers. This is a Good Thing.

Comment (lmorchard) [inline docs/serving-static-content.md]: > I don't think it's always safe to assume that apps can, should, or will be good at serving their own static files.

Yeah, but that's why we have CDNs and caching proxies. I think it simplifies things to encourage the server-side app to host its own static files *at least as an origin* to some other intermediary upstream, when necessary.

**Comment by lmorchard** [comment]
> A build process that includes compiling static files to a predetermined directory and pushing them to S3.

I'd actively discourage this as a way to deploy static assets that are tied to a server-side app encapsulated as a Docker container. It adds more artifacts in different places to be managed. If you need better performance, then a proxy with a flushable / expirable cache would be easier to manage than an S3 bucket IMO.

**Comment by mostlygeek** [comment]
> This covers some use cases (such as ours) fine, but if we're looking to...

Good points. I think I will change the recommendation so serving from the app is the first choice however the hosting provider can do whatever they want, like copy out assets. Though it is a discouraged it will open the door for very specific custom use cases.

**Pull request merged**